### PR TITLE
Bring back and finish image loading optimization

### PR DESCRIPTION
**Pull request by alpaix**
This change is to bring back image loading optimization while revisiting its implementation.

Key changes are:
- Use of MemoryCache instead of a regular dictionary.
- Image loading logic moved from a model to a value convertor.
- Use of shared style for package icon images.
- Code factorization to allow further re-use.
- Sliding error threshold

Fixes https://github.com/NuGet/Home/issues/2187.

//cc @yishaigalatzer @emgarten @rrelyea

### CODE REVIEW

File: src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+{
+    /// <summary>
+    /// Sliding in time error threshold evaluator. 
+    /// Verifies the number of fatal errors within last hour not bigger than reasonable threshold.
+    /// Not thread-safe as all value converter calls happen in main UI thread.
+    /// </summary>
+    internal class ErrorFloodGate
+    {
+        // If we fail at least this high (failures/attempts), we'll shut off image loads.
+        // TODO: Should we allow this to be overridden in nuget.config.
```
Comment (rrelyea) [inline src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs:14]: lets decide if this is necessary or not...and kill the todo.

File: src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        public void ReportError()
+        {
+            int ticks = GetTicks(_origin);
+            _failures.Enqueue(ticks);
+        }
+
+        private static int GetTicks(DateTimeOffset origin)
+        {
+            return (int)((DateTimeOffset.Now - origin).TotalSeconds / 5);
```
Comment (rrelyea) [inline src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs]: never suspected that a tick was 5 seconds. made the code hard to interpret in IsOpen.
combination of comments and/or method name renaming might be good.

File: src/NuGet.Clients/PackageManagement.UI/Resources/Resources.xaml [hunk 3]
```diff
@@ -135,27 +135,23 @@
...
+    <Setter Property="Source">
+      <Setter.Value>
+        <Binding
+          Path="DataContext.IconUrl"
+          RelativeSource="{RelativeSource Self}"
+          Converter="{StaticResource IconUrlToImageCacheConverter}"
+          ConverterParameter="{x:Static nuget:Images.DefaultPackageIcon}"
+          TargetNullValue="{x:Static nuget:Images.DefaultPackageIcon}"
+          FallbackValue="{x:Static nuget:Images.DefaultPackageIcon}"
+          IsAsync="False"/>
```
Comment (rrelyea) [inline src/NuGet.Clients/PackageManagement.UI/Resources/Resources.xaml]: IsAsync false is default behavior. better to not set values to default values in xaml.

**Comment by rrelyea** [comment]
:shipit:

File: src/NuGet.Clients/PackageManagement.UI/Resources/Resources.xaml (hunk 3 above)
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.UI/Resources/Resources.xaml]: Agree

File: src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs (hunk 2 above)
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.UI/Common/ErrorFloodGate.cs]: You're right. Will add more comments on that.

**Pull request merged**